### DISCUSSION TITLE
Molecule/species name in surface flux units

### DISCUSSION THREAD

**Discussion by briardew** [cf-convention/discuss#350] [Comments and ideas for changing CF]
### Topic for discussion

I'm curious if CF could add a convention for indicating the molecule/species name for a surface flux. I'll use the example I work with the most, CO2. In current CF conventions, you would indicate a surface flux as `kg m-2 s-1`. However, CO2 fluxes are often provided in units of C, not units of CO2. So there is considerable ambiguity here. Indicating `kg C m-2 s-1` would be nonsensical as `C` denotes Coulomb. Would it be possible to perhaps indicate the specific molecule/species in square brackets, for example, `kg [C] m-2 s-1` or `kg [CO2] m-2 s-1`? Resolving this ambiguity is essential for Earth system modeling as the two differ by about a factor of 3.

**Comment by JonathanGregory** [comment]
Dear Brad

Thanks for your question. In CF, the `units` are SI, and they are not used to identify the physical quantity. For that purpose, CF has the `standard_name` attribute. We have standard names to distinguish different ways of expressing fluxes of chemical species. For instance, `surface_upward_mass_flux_of_ammonia` refers to the entire mass of the ammonia molecules, whereas `surface_upward_mass_flux_of_nitrogen_compounds_expressed_as_nitrogen` refers to the mass of the nitrogen only in the molecules. They have the same `units` of `kg m-2 s-1`. There are many standard names for CO2 fluxes with the unit of `kg m-2 s-1` e.g. `surface_downward_mass_flux_of_carbon_dioxide_expressed_as_carbon`. As you say, in the literature that quantity would sometimes be given units of `kgC m-2 s-1`, but that's not SI. At present, all the defined standard names for CO2 fluxes are `expressed_as_carbon`. That means no-one has yet requested e.g. `surface_downward_mass_flux_of_carbon_dioxide` as a standard name, but if it were needed it would probably be approved. Please search [the current standard name table](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html) for "carbon dioxide" to see what's currently defined.

Cheers

Jonathan

**Comment by cseaton** [comment]
Dear Jonathon,

One thing I discovered yesterday while looking into Brad's question is that the 'expressed_as' modifier does not seem to be explicitly described anywhere in the CF documentation. Specifically, I thought I would find it in the Guidelines for Construction of CF Standard Names (cfconventions.org)<https://cfconventions.org/Data/cf-standard-names/docs/guidelines.html#id2798851> and also in section 3.1 Units of the main CF convention. The description of 'expressed_as' does show up in the individual standard name definitions
The phrase "expressed_as" is used in the construction A_expressed_as_B, where B is a chemical constituent of A. It means that the quantity indicated by the standard name is calculated solely with respect to the B contained in A, neglecting all other chemical constituents of A.
but I wonder if that shouldn't be repeated somewhere in the guidelines for standard names or in the Units section of the convention.

Thanks,
Charles
________________________________

**Comment by briardew** [comment]
Hi Jonathan and Charles,

I think I still haven't done a very good job at explaining what is
unsatisfactory about the current standard.

If we're talking about carbon fluxes, there are all sorts of fluxes I may
want to read into an Earth system model or an analysis program. For
brevity, I won't bother defining these terms, many of which are
combinations of the others. For example, there is GPP, NPP, Rh, Ra, Re,
wildfires, wood fuel burning, harvest, grazing, cement carbonation,
deforestation, forest regrowth, peat drainage, peat fires, etc. etc. All of
these fluxes could reasonably be defined in terms of kg C m-2 s-1 or kg CO2
m-2 s-1. To rely upon standard_name for all of them seems like an
impossible task. There's also fictitious fluxes, test pulses, and
perturbations I may want to test and compare to other products.

I assure you the current standard is insufficient and there is debate in
the community as to how to best address it. If anything, it's most likely
that I'm not doing this problem justice.

Thanks,
Brad

**Reply by cseaton** [reply]
Hi Brad,

This problem seems similar to biological taxa or geographic regions, where it doesn't make sense to have a standard name for each species or region, just a standard name that specifies that a variable contains names/identifiers that further specify something described in a standard name.

Since these are all CO2 fluxes, they would all be covered by the standard name surface_downward_mass_flux_of_carbon_dioxide_expressed_as_carbon (or _as_carbon_dioxide), 
and what is potentially needed is a standard name for the label variable that specifies the cause of the flux. 

Listing the cause can be handled in CF using label auxiliary coordinates: https://cfconventions.org/cf-conventions/cf-conventions.html#labels. As with biological taxa (standard_name="biological_taxon_name") or geographic regions (standard_name="region"), you could propose a new standard name like "cause_of_co2_fluxes' or just 'cause_of_fluxes' or 'caused_by' so that the label variable could be programmatically identified as being that specific sort of label. The controlled vocabulary for those terms would be handled outside of CF.
 
An example CF (with the definition of the cause of flux in bold)
dimension:
  * time = 100 ;
  * string80 = 80 ;
  * **cause = 2 ;**

variables:
  * float time(time);
    * time:standard_name = "time" ;
    * time:units = "days since 2019-01-01" ;
  * float co2flux(time,cause) ;
    * co2flux:standard_name = "surface_downward_mass_flux_of_carbon_dioxide" ;
    * **co2flux:coordinates = "cause" ;**
    * units="kg m-2 s-1"
    * **units_metadata= [not currently in the standard] "expressed_as:carbon" **
  * **char cause_name(cause,string80) ;
    * cause_name:long_name = "cause of co2 flux using some external controlled vocabulary" ;**
    * cause_name:standard_name = [example new standard name] "cause of flux"

data:
  * time = // 100 values ;
  * co2flux = // 200 values ;
  * cause_name = "GPP", "grazing" ;   // names from some external controlled vocabulary

This also has the advantage that if you are specifying many different CO2 fluxes by cause in the same file, you can incorporate them into a single variable with a label coordinate that specifies the different causes.

Charles

**Comment by taylor13** [comment]
HI Brad,

Maybe, if you have a particular quantity in mind that has no standard name, you could propose a standard name, and together we might be able to find a way to make it fully descriptive.  If your proposal is to relax the CF convention's restrictions on units, that would be disruptive enough that it would undoubtedly require extended discussion.  This might be warranted if somehow the standard names were unable to handle your use case(s), but I don't think that has yet been demonstrated by a specific example. 

Karl

**Comment by JonathanGregory** [comment]
Dear Brad @briardew

It's not mandatory to use standard names in CF; they're optional. If you're concerned only with your own data and analysis, you don't need to use them. You could define your own private convention with a different attribute, or standardise the `long_name` for your own purposes, for example.

The main purpose of standard names is to facilitate sharing of data from different sources. For that purpose you need to define the quantities you are going to exchange, so you have a common understanding of what they are. As Karl says, you can propose new standard names. CF defines them for real use-cases as they arise. If CF agrees a standard name for a quantity in `kg m-2 s-1` of CO2, it would almost certainly be unproblematic to agree one containing `expressed_as_carbon` for the corresponding quantity in `kg m-2 s-1` of carbon as well, or vice-versa.

Best wishes

Jonathan

**Comment by JonathanGregory** [comment]
Dear Charles @cseaton 

The guidelines are not an exhaustive list of all the patterns and phrases that have been used in formulating standard names, so it isn't irregular for `expressed_as` not to be included. It's a good suggestion that we should include it, though, since it appears in over 300 existing standard names. Thanks for the idea.

Cheers

Jonathan

**Comment by bnlawrence** [comment]
Hi Brad

Both @cseaton and @JonathanGregory have made concrete suggestions, but if they don't work, I'd be keen to understand why, we have a lot of ESM use-cases here at NCAS, and I like to understand problems before they hit me :-).

I would say though that whatever we do, mucking with our dependence on on SI units is likely the very last choice, as it would be very disruptive; e.g. CF compliant code can currently reasonably expect to _understand_ SI units and some packages can do on the fly unit conversions between fields if the data units are not the same as (but are still consistent with) the Std Name, so adding anything new to the concept of units would have a lot of side effects.

Cheers
Bryan

**Reply by briardew** [reply]
Hi Bryan (and all),

Thanks to everyone for bearing with me while I sort of stumbled around this topic.

One thing to start off with is that this mostly just a C vs CO2 issue. I can't imagine a scenario where one might want to talk about emissions of say CFC-12 in terms of CH4 or even C. The basic issue to me is that specifying units of `kg m-2 s-1` doesn't yet uniquely specify the flux. When using `kg`, we also need to know of what. This is why we see the prevalence of standard names with and without the suffix `expressed_as_carbon`.

It seems to me like CF should specify a convention that uniquely specifies the variable. If I'm developing a GCM, I cannot read a flux file, check its units attribute, find `kg m-2 s-1` and know how to proceed. I must still know `kg` of what. Shouldn't CF define a standard for telling GCMs, software, etc. precisely what this is? Maybe an easier way would be to suggest another attribute for the variable, perhaps `expressed_as` 😄, that would uniquely specify the field?

Thanks,
Brad

**Comment by JonathanGregory** [comment]
Dear Brad

You write that "specifying units of `kg m-2 s-1` doesn't yet uniquely specify the flux." That's right, because the function of `units` is solely to state the units :smiley:! Obviously, there are _many_ quantities with `units` of `kg m-2 s-1`, not just fluxes of CO2. CF provides other metadata to describe the geophysical quantity, especially the standard name.

The standard name regards a flux of CO2 in `kg m-2 s-1` and a flux of CO2 expressed as `kg m-2 s-1` of carbon as different quantities. The difference is similar to that between `thickness_of_snowfall_amount` and `lwe_thickness_of_snowfall_amount`, which are both in `m`. LWE stands for liquid water equivalent. These quantities are the actual thickness of show that has fallen, and the thickness of water you get by melting the snow. They're two ways of expressing the same information, but with the same unit. This information can also be expressed as `snowfall_amount`, which is in a different unit, `kg m-2`.

Although there are many CO2 fluxes expressed as carbon, it's not unique to CO2. For instance, we have `surface_upward_mass_flux_of_nitrous_oxide_expressed_as_nitrogen_out_of_vegetation_and_litter_and_soil`.

Best wishes

Jonathan
